### ISSUE TITLE
[Feature Request] Bring back the mask option in SDPA for GGML compatiablity

### ISSUE THREAD

**Issue by marty1885** [tenstorrent/tt-metal#10376]
**Is your feature request related to a problem? Please describe.**
I noticed in https://github.com/tenstorrent/tt-metal/pull/10313/files#diff-a97ed6437b65106019f0555fd396e7919b7c2c50b04f9b03213c9cdd8bb932a1 the `mask` parameter is replaced `curr_pos`. GGML uses the FlashAttention-v1 format for it's `GGML_OP_FLASH_ATTN_EXT` operator.  This change breaks my ability to hook FlashAttention into GGML (I haven't hooked it up due to other issues, but this change completely breaks it).

```c
struct ggml_tensor * ggml_flash_attn_ext(
        struct ggml_context * ctx,
        struct ggml_tensor  * q,
        struct ggml_tensor  * k,
        struct ggml_tensor  * v,
        struct ggml_tensor  * mask,
        float                 scale,
        float                 max_bias) {
    GGML_ASSERT(ggml_can_mul_mat(k, q));
    // TODO: check if vT can be multiplied by (k*qT)

    if (mask) {
        GGML_ASSERT(ggml_is_contiguous(mask));
        GGML_ASSERT(mask->ne[2] == 1);
        GGML_ASSERT(mask->ne[3] == 1);
        GGML_ASSERT(mask->ne[1] >= GGML_PAD(q->ne[1], GGML_KQ_MASK_PAD) &&
                "the Flash-Attention kernel requires the mask to be padded to GGML_KQ_MASK_PAD and at least n_queries big");
        //GGML_ASSERT(ggml_can_repeat_rows(mask, qk));
    }

   // ...... non-important code

    result->op   = GGML_OP_FLASH_ATTN_EXT;
    result->src[0] = q;
    result->src[1] = k;
    result->src[2] = v;
    result->src[3] = mask; // GGML expects mask as a tensor

    return result;
}
```

**Describe the solution you'd like**

Bring back the mask parameter for SDPA

**Describe alternatives you've considered**

**Additional context**

**Comment by cglagovichTT** [comment]
@marty1885 thanks for the feedback. We should be able to change the op to support `mask` or `cur_pos` and choose functionality based on which is provided.

Just to confirm, this is for GGML's FlashAttention in decode mode? where
```python
Q: [1, b, nh, dh]
K,V: [b, nh, kv_len, dh]
mask: [1, b, nh, dh]
```

While we're on the topic, do you mind describing the other issues which make our FlashAttention and FlashDecode operations incompatible with GGML?

**Comment by marty1885** [comment]
GGML expects something slightly different (note: if you read the [comment in code](https://github.com/ggerganov/llama.cpp/blob/1bdd8ae19f50bc6f108fa247e90688e5c60559fc/ggml/include/ggml.h#L1741-L1754), it's reversed. GGML stores shapes in reverse order. And by default does assumes transposed MatMul to be the norm).

```
q:    [1, n_head,    n_batch,     n_embd]
k:    [1, n_head_kv, n_kv,        n_embd]
v:    [1, n_head_kv, n_kv,        n_embd] !! not transposed !!
mask: [1, 1,         n_batch_pad, n_kv]   !! n_batch_pad = GGML_PAD(n_batch, GGML_KQ_MASK_PAD) !!
res:  [1, n_batch,   n_head,      n_embd] !! permuted !!

GGML_KQ_MASK_PAD = 32 by default. It matches nicely with TT hardware
```

Currently I'm handling it by calling `transpose` and `repeat` manually to convet. But it hangs often and I'm debugging what's going on (the issue I mentioned above), unsure if it's bug in my code.

I need to check the differences, but the top things on my head:

* K, V does not support batches
* The mask is broadcasted (and optional)
* Query transposed
* Supports ALiBi (`max_bias == 0` disables it)
* Does not need the `valid_seq_len` parameter

It would be great if TT can add ALiBi support also to it's softmax operation. I was waiting to report performance related missing features after I can bring up an LLM.

**Comment by caixunshiren** [comment]
Hey @marty1885, thanks for bringing this up.

I can add the support for attention mask. As an initial quick fix to unblock your work, I can bring the legacy support back for v1. However, v1 would require `padded_sequence_length` and the attention mask is assumed to be `[1, b, nh, padded_sequence_length]`. It appears to me that the mask you your ggml code (`[1, 1,         n_batch_pad, n_kv]`) is a different kind of mask that mask out batch and kv head. 

On the other hand, the purpose of the mask in flash decode v1 is due to an inefficient implementation, which we do attention on the longest sequence length of the batch, and use mask to mask out the redundant computation on shorter sequence length. It is completely unnecessary to do that, so that's why we got rid of attention mask and use start pos instead.

Is the v1 mask functionality really what you intend to use?

**Comment by marty1885** [comment]
@caixunshiren Good question. Will reply soon. I need some time to completely confirm exactly.

**Comment by ayerofieiev-tt** [comment]
@marty1885, should I close this now?

**Comment by marty1885** [comment]
@ayerofieiev-tt I still need this. I haven't got my backend back into a healthy enough stage where I can explore this further. Largely limited by the following issues

* https://github.com/tenstorrent/tt-metal/issues/11650
* https://github.com/tenstorrent/tt-metal/issues/11082

Please let me know your thoughts.

**Comment by minaliuTT** [comment]
@cglagovichTT Just checking, is this issue still valid, or can it be closed?

**Comment by marty1885** [comment]
Please keep this. Assigning to myself as I am writing one.
